adding recipe for StopWatch

dcatt-haiku opened the pull request:
My first attempt at creating a recipe for my application.  I am in learning mode for this, so I apologize in advance for any mistakes unrealized.

bbjimmy commented:
I took a look at your recipe and found several issues.

- The stopwatch repo has no makefile

- Usually one does a release on the source repository to give the recipe a secure source. 

- The recipe name must be all lower case.

I suggest installing haikuporter and attempting to run the recipe file. Once it runs and builds an .hpkg file, then push your changes to your haikuports repo to update this PR.

humdingerb commented:
Yep, heed bbjimmy's advice: do try to build your recipe with haikuporter. :)
I see that you have remnants of your recipe "template" quicklaunch still in there. :)   
The "make bindcatalogs" is only needed if your app is localized and provides catalogs.   
The ARCHITECTURES should only list those you've tested on, i.e. "x86_gcc2  x86". Where it's unknown, you can put a "?" in front of, e.g. "?x86_64". Where it doesn't work a "!".
For a secure SOURCE_URI, refer back to quicklaunch. Then it needs a CHECKSUM_SHA256 as well.

Also have a look at the [wiki](https://github.com/haikuports/haikuports/wiki/DevelopmentModel) esp. "How do I make one final commit from many for the pull request (PR)?"

dcatt-haiku commented:
BBimmy, Humdinger, 

A few questions...

1) I've never create a makefile before, so if you have a any direction on that that would be great
2) Yes, I did notice the SHA bit, but couldn't find where to generate that to include in the SOURCE_URI
3) I had issues with the errors on the ARCHITECTURES bit even though I had my haikuports.conf inline with my recipe, so I changed it to be less restrictive

And sorry about the remnant, I thought I had it scanned and updated pretty well.  Even gave it one last scene before the commit.  I fix that along with some of this other stuff.

humdingerb commented:
1) The Paladin makefile might work. Otherwise, there's the makefile_engine "makefile" template at /boot/system/develop/etc/Makefile . I think it's also available from Tracker's "New" menu.

2) You either use "sha256sum" on the archive that's downloaded by haikuporter or, easier, just put an empty CHECKSUM_SHA256="" in the recipe and haikuporter will tell you what to enter.

3) ARCHITECTURES="x86_gcc2" should work with an official nightly (gcc2hybrid).

After you cloned haikuporter and forked/cloned haikuports, you'll have to put your new recipe into the tree, e.g. in haikuports/haiku-apps/stopwatch/.
You call haikuporter with "haikuporter --get-dependencies -s stopwatch". (Look into the wiki how to edit the "profile" so you can shortcut that to "hp stopwatch".
Hope that helped.

dcatt-haiku commented:
So I've satisfied pretty much everything now, but when I build my recipe, it returns this error on soemthing I have no clue about...

 haikuporter -S stopwatch
Checking if any package-infos need to be updated ...
Looking for stale dependency-infos ...
======================================================================
haiku-apps::stopwatch-1.0.0
======================================================================
The following required ports will be built first:
        dev-libs::libffi-3.0.13
        sys-devel::patch-2.7.5
        sys-devel::make-4.1
        sys-apps::sed-4.2.1
        sys-apps::diffutils-3.3
        app-arch::tar-1.28
        sys-apps::help2man-1.47.3
        sys-libs::ncurses-5.9
        sys-devel::m4-1.4.17
        dev-lang::perl-5.18.2
        sys-libs::ncurses6-6.0
        dev-libs::expat-2.1.1
        sys-libs::zlib-1.2.8
        app-arch::bzip2-1.0.6
        dev-libs::libiconv-1.13.1
        sys-apps::groff-1.20.1
        dev-libs::libpcre-8.38
        sys-apps::less-481
        dev-db::sqlite-3.12.2.0
        sys-apps::coreutils-8.24
        sys-apps::findutils-4.4.2
        sys-apps::texinfo4-4.13a
        sys-libs::readline-6.3.8
        dev-libs::openssl-1.1.0~git
        app-arch::gzip-1.8
        dev-util::cmake-2.8.11.2
        sys-devel::gcc-2.95.3_2014_10_14
        sys-devel::binutils-2.17_2013_04_21
        sys-apps::gawk-4.1.3
        app-editors::nano-2.5.3
        sys-devel::gettext-0.19.7
        sys-devel::libtool-2.4.2
        app-shells::bash-4.3.42
        sys-devel::autoconf-2.69
        sys-devel::bison-3.0.4
        net-misc::curl-7.48.0
        sys-apps::grep-2.24
        sys-devel::flex-2.6.0
        sys-apps::texinfo-6.1
        dev-util::pkgconfig-0.29.1
        sys-devel::automake-1.15.0
        dev-lang::python-2.7.11
        sys-apps::man-1.6g
        dev-vcs::git-2.8.0
----------------------------------------------------------------------
dev-libs::libffi-3.0.13
        /boot/home/projects/GitHub/dcatt-haiku/haikuports/dev-libs/libffi/libffi-3.0.13.recipe
----------------------------------------------------------------------
Skipping download of source for libffi-3.0.13.tar.gz
Skipping checksum validation of libffi-3.0.13.tar.gz
Skipping unpack of libffi-3.0.13.tar.gz
Skipping patchset for libffi-3.0.13.tar.gz
Cleaning up temporary directories ...
[Errno -2147459069] No such file or directory: '/boot/home/projects/GitHub/dcatt-haiku/haikuports/repository/libffi_source-3.0.13.DependencyInfo'

bbjimmy commented:
looks like you need :

SOURCE_DIR=StopWatch-1.0.0

make it the next line after the checksum.

dcatt-haiku commented:
So I did what you said, which got me farther, but now it fails on build with this and I troubleshooted it with various approaches and can't seem to make it happy short of symlinking makefile-engine into /etc (if that is what it wants?);

waiting for build package stopwatch-1.0.0-1 to be activated
Building ...
Makefile:27: /etc/makefile-engine: No such file or directory
make: *** No rule to make target '/etc/makefile-engine'.  Stop.

bbjimmy commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+BUILD_PREREQUIRES="
+	cmd:g++
+	cmd:make
+	"
```
need to add makefile_engine

bbjimmy commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+
+BUILD()
+{
+	make $jobArgs OBJ_DIR=objects
```
add:
rc ./Resources.rdef

bbjimmy replied:
make $jobArgs BUILDHOME=`finddir B_SYSTEM_DEVELOP_DIRECTORY`

humdingerb replied:
If you use the Makefile template of the makefile-engine, there's no need to convert rdef to rsrc. There's also no need to define a BUILDHOME with it.

bbjimmy commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+	stopwatchDir=$appsDir/StopWatch
+	mkdir -p $stopwatchDir
+
+	cp -af objects/StopWatch $stopwatchDir
```
shoukd be:

 cp -af objects.*/StopWatch $stopwatchDir

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+SUMMARY="A simple stopwatch application"
+DESCRIPTION="A simple stopwatch application."
```
Maybe you can come up with a little more detailed description. :)

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+ARCHITECTURES="x86_gcc2 ?x86 ?x86_64"
+
+PROVIDES="
+	stopWatch = $portVersion
```
should be "stopwatch = $portVersion"

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,42 @@
+
+INSTALL()
+{
+	stopwatchDir=$appsDir/StopWatch
```
If you only have your one binary and no other files, like documentation, there's no need to put it in a subfolder.

dcatt-haiku commented:
Okay, made more changes to to the recipe based on comments, but running the recipe still fails...

`waiting for build package stopwatch-1.0.0-1 to be activated
Building ...
Makefile:27: /etc/makefile-engine: No such file or directory
make: Warning: File 'Makefile' has modification time 11151 s in the future
make: *** No rule to make target '/etc/makefile-engine'.  Stop.
Command '['bash', '-c', '. /wrapper-script']' returned non-zero exit status 2
keeping chroot folder /boot/home/projects/GitHub/dcatt-haiku/haikuports/haiku-apps/stopwatch/work-1.0.0 intact for inspection
Error: Build has failed - stopping.`

dcatt-haiku commented:
Tried yet another approach mimicking the CDPlayer Makefile and Recipe somewhat hoping for success but it still fails on the lack of finding the makefile-engine and I actually ran the CDPlayer recipe with success locally and figured maybe I'd have success with this recipe.

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,40 @@
+ARCHITECTURES="x86_gcc2 ?x86 ?x86_64"
+
+PROVIDES="
+	StopWatch = $portVersion
```
All lower-case, like the recipe name: "stopwatch = $portVersion"

The "app:StopWatch = $portVersion" below is correct. This line declares the actual filename of the application.

dcatt-haiku replied:
I think I forgot to change that back as I was desperately trying anything to get the recipe to work and was taking ideas from other recipes.  Would you be willing to pull this down and try to run it and see if you can replicate the issue I'm seeing with the work-* directory?

humdingerb replied:
I just did, and (with that last change "cp -af objects/StopWatch $appsDir"), it does build.
So, the problem seems to be with your configuration. What's the output of "findpaths -r "makefile_engine" B_FIND_PATH_DEVELOP_DIRECTORY"? It should be "/boot/system/develop". If it isn't maybe that package isn't correctly installed after all. Try "pkgman search makefile_engine" to see if it's installed (then it's marked with an "s"), and maybe reinstall...

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,40 @@
+INSTALL()
+{
+	mkdir -p $appsDir
+	cp -af objects.*/StopWatch $appsDir
```
Since you use "make $jobArgs OBJ_DIR=objects" to build, this like should be "cp -af objects/StopWatch $appsDir"

dcatt-haiku replied:
BBJimmy suggested what you see above.  So I'm just trying to get this thing as correct as possible.

humdingerb replied:
Yep, which would be correct if you didn't use "OBJ_DIR=objects " in your make line. That makes all objects and the executable appear in a "objects" folder instead of one with the used compiler and debug-status in its name, e.g. "objects.debug.x86_gcc2" or something. Then you'd need the wildcard "*" to mask that suffix.   
Found your makefile-engine yet? :)

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,40 @@
+COPYRIGHT="2016 DCatt"
+LICENSE="MIT"
+REVISION="1"
+SOURCE_URI="https://github.com/dcatt-haiku/StopWatch/archive/v1.0.0.tar.gz"
```
You could change this to "https://github.com/dcatt-haiku/StopWatch/archive/v$portVersion.tar.gz" to make updates to future release even easier.

humdingerb commented on haiku-apps/stopwatch/stopwatch-1.0.0.recipe:
```diff
@@ -0,0 +1,40 @@
+REVISION="1"
+SOURCE_URI="https://github.com/dcatt-haiku/StopWatch/archive/v1.0.0.tar.gz"
+CHECKSUM_SHA256="320f0d7502316e12209c73576cacf488f49abd58a42cc986a84882a54eb9e732"
+SOURCE_DIR="StopWatch-1.0.0"
```
Same here: "SOURCE_DIR="StopWatch-$portVersion"

humdingerb commented:
Glad you made it! Now you'll just have to squash your 15 commits to get there. :)
See [How do I make one final commit from many for the pull request (PR)?](https://github.com/haikuports/haikuports/wiki/DevelopmentModel) at the buttom of the page.

dcatt-haiku commented:
humdinger wrote:
>
> Glad you made it! Now you'll just have to squash your 15 commits to 
> get there. :)
> See How do I make one final commit from many for the pull request 
> (PR)? <https://github.com/haikuports/haikuports/wiki/DevelopmentModel> 
> at the buttom of the page.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/haikuports/haikuports/pull/580#issuecomment-216808659>
>
I can't get the final commit to work.  I tried the three options but 
with no local changes to invoke a commit to push is leaving me dead in 
the water.  I'll keep messing with it and see if I can get it in.

fbrosson commented:
This _might_ work, but maybe it's safer to wait for a confirmation from a git expert.
```
git pull upstream master
git rebase --interactive upstream/master
# the preivous line will get you into vim or $EDITOR
git push
```
The second command should start an editor with a list of all your commit titles. Keep the first line as **`pick`** and change the next ones to either **`squash`** or **`fix`**, depending on whether or not you want its commit comment to be copied to the  "commit comments edit session" that will follow. I guess the only tricky step was the **`git rebase --interactive upstream/master`**.
BTW, if you prefer vim over nano then you need to **`export EDITOR=vim`**.

waddlesplash commented:
> Since I was working from master to begin with, maybe those instructions won't work. Shall I close this PR and start over?

Please don't, and instead manually squash your changes (or do a fresh clone, manually re-commit your work, and then push --force.)

fbrosson commented:
@dcatt-haiku, if you wish to try a squash you can do it safely on a branch and then, once you have noticed that everything works as expected you can redo it in your master (for this time only, because as @waddlesplash wrote and as you already know, it is not recommended to commit directly to one's master when doing PRs).
So here are the steps you might try:

* Do a backup of the recipe in a safe place, just in case something goes wrong.
* Check with "**`git remote -v`**" that https://github.com/haikuports/haikuports.git is listed as one (or more) of your tracked repositories. I think haikuporter automatically adds it as "real_origin" but you can add it as "upstream" if you wish. In case you don't already have it, you can:
```
git remote add upstream https://github.com/haikuports/haikuports.git
```

* Create a test1 branch:
```
git checkout -b test1 master
```

* Try a squah in your test1 branch, replacing **`upstream`** by **`real_origin`** (and **`upstream/master`** by **`real_origin/master`**) if you already had **`real_origin`** and chose not to add **`upstream`**:
```
git pull upstream master
git rebase --interactive upstream/master
```

* Check with **`git log`** and/or **`git diff upstream/master`** that your test1 branch only has one extra commit compared to upstream/master. If that's the case then you can redo the squash on your master (and push your master to your GitHub repo) with these steps:
```
chekout master
git pull upstream master
git rebase --interactive upstream/master
git push -f
```

* If your test1 branch went wrong you can:
```
git checkout master
git branch -D test1
```

and retry. Of course, you can keep test1 and retry with test2...

* Once your PR is merged then you should probably redo a fresh clone and leave master clean ;-)

In case you are not confortable with the squash method, then you can try the "fresh clone + manual re-commit" method but in that case I don't know how you would clone from a clean master instead of your already modified master on GitHub and then make your master track your clone on GitHub.
@waddlesplash or anyone else: please correct me if I'm mistaken. Thanks!

waddlesplash commented:
> In case you are not comfortable with the squash method, then you can try the "fresh clone + manual re-commit" method but in that case I don't know how you would clone from a clean master instead of your already modified master on GitHub and then make your master track your clone on GitHub.

```
git clone https://github.com/haikuports/haikuports.git
<re-commit your changes>
git push --force https://github.com/your_name_here/haikuports.git
```

dcatt-haiku commented:
Sorry for the radio silence guys... been busy as of late.  I'll try to tackle this by the end of the weekend (if not before).

dcatt-haiku commented:
Sorry for the radio silence guys... been busy as of late.  I'll try to tackle this by the end of the weekend (if not before).

dcatt-haiku commented:
Closed pull request to start over fresh.

korli commented:
A kind reminder: github can squash the changes itself now.
